### PR TITLE
Internal: Editor V2 - Site Settings [ED-9727]

### PR DESCRIPTION
**Pull request by Nevoss**

### CODE REVIEW

File: packages/.eslintrc.js [hunk 1]
```diff
@@ -58,12 +58,17 @@ module.exports = {
...
 				'**/@(__mocks__|__tests__|tests|test)/**/*.[tj]s?(x)', // Test files.
 			],
 			rules: {
 				// In tests, we are importing dev dependencies of the root directory, so we need to disable this rule.
 				'import/no-extraneous-dependencies': [ 'off' ],
+				'import/no-unresolved': [ 'error', {
+					ignore: [
+						'^test-utils$', // In tests, it should ignore the test-utils helper.
+					],
+				} ],
```
Comment (StyleShit) [inline packages/.eslintrc.js:71]: do you wanna move it to another PR so we won't have 30 files here?

File: packages/packages/documents/src/init.ts [hunk 2]
```diff
@@ -23,7 +23,7 @@ function registerTopBarMenuItems() {
 	} );
 
 	injectIntoPrimaryAction( {
-		name: 'primary-action',
+		name: 'top-bar-primary-action',
```
Comment (StyleShit) [inline packages/packages/documents/src/init.ts:26]: shouldn't be "document-primary-action"?

File: packages/packages/site-settings/src/components/__tests__/ported-primary-action.test.tsx [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+	el.setAttribute( 'id', id );
+
+	document.body.appendChild( el );
+}
+
+function navigateTo( route: string ) {
+	window.dispatchEvent( new CustomEvent( 'elementor/routes/open', {
+		detail: { route },
+	} ) );
+}
```
Comment (StyleShit) [inline packages/packages/site-settings/src/components/__tests__/ported-primary-action.test.tsx:93]: maybe we need to move it to the test utils, and use it everywhere, and also the command utils I created

File: packages/packages/site-settings/src/components/__tests__/ported-primary-action.test.tsx [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+		jest.mocked( useActiveDocument ).mockImplementation( () => createMockDocument() );
+	} );
+
+	afterEach( () => {
+		jest.clearAllMocks();
+
+		document.body.innerHTML = '';
+	} );
+
+	it( 'should render the button when navigating to site-settings route', () => {
```
Comment (StyleShit) [inline packages/packages/site-settings/src/components/__tests__/ported-primary-action.test.tsx:30]:
```suggestion
	it( 'should render the button when opening the site-settings panel', () => {
```

be less technical

File: packages/packages/site-settings/src/components/__tests__/ported-primary-action.test.tsx [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+		// Assert.
+		expect( screen.queryByRole( 'button' ) ).not.toBeInTheDocument();
+
+		// Act.
+		jest.mocked( isRouteActive ).mockImplementation( () => true );
+
+		navigateTo( 'panel/global/menu' );
+
+		// Assert.
+		expect( screen.queryByRole( 'button' ) ).toBeInTheDocument();
```
Comment (StyleShit) [inline packages/packages/site-settings/src/components/__tests__/ported-primary-action.test.tsx:45]: maybe also `getByText()` or something?

File: packages/packages/site-settings/src/components/__tests__/ported-primary-action.test.tsx [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+	} );
+
+	it( 'should not render button when navigation to another route', () => {
```
Comment (StyleShit) [inline packages/packages/site-settings/src/components/__tests__/ported-primary-action.test.tsx:48]:
```suggestion
	it( 'should not render the button when opening another panel', () => {
```

File: packages/packages/site-settings/src/components/__tests__/ported-primary-action.test.tsx [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+		jest.mocked( isRouteActive ).mockImplementation( () => false );
+
+		render( <PortedPrimaryAction /> );
+
+		// Assert.
+		expect( screen.queryByRole( 'button' ) ).not.toBeInTheDocument();
+
+		// Act.
+		jest.mocked( isRouteActive ).mockImplementation( () => true );
```
Comment (StyleShit) [inline packages/packages/site-settings/src/components/__tests__/ported-primary-action.test.tsx:60]: why do you return true?

File: packages/packages/site-settings/src/components/__tests__/ported-primary-action.test.tsx [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+		navigateTo( 'panel/history' );
+
+		// Assert.
+		expect( screen.queryByRole( 'button' ) ).not.toBeInTheDocument();
+	} );
+
+	it( 'should not render button when root element not exists', () => {
```
Comment (StyleShit) [inline packages/packages/site-settings/src/components/__tests__/ported-primary-action.test.tsx:68]:
```suggestion
	it( 'should not render the button when the root element does not exist', () => {
```

File: packages/packages/site-settings/src/components/__tests__/ported-primary-action.test.tsx [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+} ) );
+
+jest.mock( '@elementor/documents', () => ( {
+	useActiveDocument: jest.fn(),
+	useActiveDocumentActions: jest.fn( () => ( {
+		save: jest.fn(),
+	} ) ),
+} ) );
+
+describe( '@elementor/documents - Site settings ported primary action', () => {
```
Comment (StyleShit) [inline packages/packages/site-settings/src/components/__tests__/ported-primary-action.test.tsx:19]:
```suggestion
describe( '@elementor/site-settings - Portalled primary action', () => {
```

File: packages/packages/site-settings/src/components/__tests__/primary-action.test.tsx [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+import PrimaryAction from '../primary-action';
+import { createMockDocument } from 'test-utils';
+import { useActiveDocument, useActiveDocumentActions } from '@elementor/documents';
+
+jest.mock( '@elementor/documents', () => ( {
+	useActiveDocument: jest.fn(),
+	useActiveDocumentActions: jest.fn(),
+} ) );
+
+describe( '@elementor/documents - Site settings primary action', () => {
```
Comment (StyleShit) [inline packages/packages/site-settings/src/components/__tests__/primary-action.test.tsx:11]:
```suggestion
describe( '@elementor/site-settings - Primary action', () => {
```

File: packages/packages/site-settings/src/components/__tests__/primary-action.test.tsx [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+		},
+		{
+			title: 'document in saving mode',
+			document: { ...createMockDocument(), isDirty: true, isSaving: true },
+		},
+		{
+			title: 'document not exists',
+			document: null,
+		},
+	] )( 'should save the document when $title', ( { document } ) => {
```
Comment (StyleShit) [inline packages/packages/site-settings/src/components/__tests__/primary-action.test.tsx:54]:
```suggestion
	] )( 'should not save the document when $title', ( { document } ) => {
```

File: packages/packages/site-settings/src/components/__tests__/primary-action.test.tsx [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+			document: { ...createMockDocument(), isDirty: false },
+		},
+		{
+			title: 'document in saving mode',
+			document: { ...createMockDocument(), isDirty: true, isSaving: true },
+		},
+		{
+			title: 'document not exists',
+			document: null,
+		},
```
Comment (StyleShit) [inline packages/packages/site-settings/src/components/__tests__/primary-action.test.tsx:53]:
```suggestion
		{
			title: "it's pristine",
			document: { ...createMockDocument(), isDirty: false },
		},
		{
			title: "it's in saving mode",
			document: { ...createMockDocument(), isDirty: true, isSaving: true },
		},
		{
			title: "it doesn't exist",
			document: null,
		},
```

File: packages/packages/site-settings/src/components/__tests__/primary-action.test.tsx [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+	it.each( [
+		{
+			title: 'document is pristine',
+			document: { ...createMockDocument(), isDirty: false },
+		},
+		{
+			title: 'document not exists',
+			document: null,
+		},
+	] )( 'should make the button disabled when $title', ( { document } ) => {
```
Comment (StyleShit) [inline packages/packages/site-settings/src/components/__tests__/primary-action.test.tsx:76]:
```suggestion
	] )( 'should be disabled when $title', ( { document } ) => {
```

File: packages/packages/site-settings/src/components/__tests__/primary-action.test.tsx [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+		// Act.
+		screen.getByRole( 'button' ).click();
+
+		// Assert.
+		expect( saveFn ).not.toHaveBeenCalled();
+	} );
+
+	it.each( [
+		{
+			title: 'document is pristine',
```
Comment (StyleShit) [inline packages/packages/site-settings/src/components/__tests__/primary-action.test.tsx:69]:
```suggestion
			title: 'the document is pristine',
```

File: packages/packages/site-settings/src/components/__tests__/primary-action.test.tsx [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+			document: { ...createMockDocument(), isDirty: false },
+		},
+		{
+			title: 'document not exists',
```
Comment (StyleShit) [inline packages/packages/site-settings/src/components/__tests__/primary-action.test.tsx:73]:
```suggestion
			title: "the document doesn't exist",
```

File: packages/packages/site-settings/src/components/__tests__/primary-action.test.tsx [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+		// Arrange.
+		jest.mocked( useActiveDocument ).mockImplementation( () => document );
+
+		// Act.
+		render( <PrimaryAction /> );
+
+		// Assert.
+		expect( screen.getByRole( 'button' ) ).toBeDisabled();
+	} );
+} );
```
Comment (StyleShit) [inline packages/packages/site-settings/src/components/__tests__/primary-action.test.tsx:86]: maybe test the loader?

File: packages/packages/site-settings/src/components/portal.tsx [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+
+export default function Portal( { children }: PropsWithChildren<object> ) {
+	const containerRef = useListenTo( [
+		routeOpenEvent( 'panel/global' ),
+		routeCloseEvent( 'panel/global' ),
+	], () => {
+		return isRouteActive( 'panel/global' )
+			? { current: document.querySelector( '#elementor-panel-inner' ) }
+			: { current: null };
+	} );
```
Comment (StyleShit) [inline packages/packages/site-settings/src/components/portal.tsx:13]: 🔥 
Nice! I would've never thought about using it like this

Maybe extract to multiple variables for readability?

File: packages/packages/site-settings/src/components/ported-primary-action.tsx [hunk 18]
```diff
@@ -0,0 +1,10 @@
+import Portal from './portal';
+import PrimaryAction from './primary-action';
+
+export default function PortedPrimaryAction() {
```
Comment (StyleShit) [inline packages/packages/site-settings/src/components/ported-primary-action.tsx:4]:
```suggestion
export default function PortalledPrimaryAction() {
```
also in other places, and also the file name

File: packages/packages/site-settings/src/components/primary-action.tsx [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+			left: 0,
+			px: 5,
+			py: 4,
+			borderTop: 1,
+			borderColor: 'grey.200',
+			background: 'white',
+		} }>
+			<Button
+				variant="contained"
+				disabled={ ! document || ! document.isDirty }
```
Comment (StyleShit) [inline packages/packages/site-settings/src/components/primary-action.tsx:23]: maybe don't render at all when there is no document?

File: packages/packages/site-settings/src/components/portal.tsx [hunk 20]
```diff
@@ -0,0 +1,24 @@
+import { PropsWithChildren } from 'react';
+import { Portal as BasePortal } from '@elementor/ui';
+import { isRouteActive, routeCloseEvent, routeOpenEvent, useListenTo } from '@elementor/v1-adapters';
+
+export default function Portal( { children }: PropsWithChildren<object> ) {
```
Comment (StyleShit) [inline packages/packages/site-settings/src/components/portal.tsx:5]:
```suggestion
export default function Portal( { children }: { children: ReactNode } ) {
```
🤷

File: packages/packages/site-settings/src/hooks/__tests__/use-action-props.test.tsx [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+} ) );
+
+describe( '@elementor/site-settings - useActionProps', () => {
+	afterEach( () => {
+		jest.clearAllMocks();
+	} );
+
+	it.each( [
+		{ title: 'open', input: { isActive: false, isBlocked: false }, expectedRoute: 'panel/global/open' },
+		{ title: 'close', input: { isActive: true, isBlocked: false }, expectedRoute: 'panel/global/close' },
```
Comment (StyleShit) [inline packages/packages/site-settings/src/hooks/__tests__/use-action-props.test.tsx:17]: maybe split to multiple lines? kinda hard to read

File: packages/packages/site-settings/src/hooks/__tests__/use-action-props.test.tsx [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+	] )( 'should $title the site-settings state when clicked', ( { input, expectedRoute } ) => {
```
Comment (StyleShit) [inline packages/packages/site-settings/src/hooks/__tests__/use-action-props.test.tsx:18]:
```suggestion
	] )( 'should $action the site-settings panel on click', ( { input, expectedRoute } ) => {
```

File: packages/packages/site-settings/src/hooks/__tests__/use-action-props.test.tsx [hunk 23]
```diff
@@ -0,0 +1,45 @@
...
+		jest.mocked( useRouteStatus ).mockImplementation( () => input );
+
+		const { result } = renderHook( () => useActionProps() );
+
+		// Act.
+		result.current.onClick();
+
+		// Assert.
+		expect( runCommand ).toHaveBeenCalledTimes( 1 );
+		expect( runCommand ).toHaveBeenCalledWith( expectedRoute );
```
Comment (StyleShit) [inline packages/packages/site-settings/src/hooks/__tests__/use-action-props.test.tsx:29]:
```suggestion
		expect( runCommand ).toHaveBeenCalledWith( expectedCommand );
```

File: packages/packages/site-settings/src/hooks/use-action-props.tsx [hunk 24]
```diff
@@ -0,0 +1,26 @@
+// import { PlusIcon } from '../icons/plus-icon';
```
Comment (StyleShit) [inline packages/packages/site-settings/src/hooks/use-action-props.tsx:1]: ?

File: packages/packages/site-settings/src/hooks/use-action-props.tsx [hunk 25]
```diff
@@ -0,0 +1,26 @@
...
+		title: __( 'Site settings', 'elementor' ),
+		icon: SettingsIcon,
+		onClick: () => {
+			if ( isActive ) {
+				runCommand( 'panel/global/close' );
+
+				return;
+			}
+
+			runCommand( 'panel/global/open' );
```
Comment (StyleShit) [inline packages/packages/site-settings/src/hooks/use-action-props.tsx:21]:
```suggestion
			isActive
				? runCommand( 'panel/global/close' )
				: runCommand( 'panel/global/open' );
```

maybe?
IDK

File: packages/packages/structure/src/hooks/use-action-props.tsx [hunk 26]
```diff
@@ -1,4 +1,3 @@
-import * as React from 'react';
```
Comment (StyleShit) [inline packages/packages/structure/src/hooks/use-action-props.tsx:1]: You can also change it to ts instead of tsx
but it's not related to this PR, so maybe extract this and the things before to another PR?

**Review by StyleShit — CHANGES_REQUESTED** [review]

File: packages/packages/documents/src/init.ts (hunk 2 above)
Comment (Nevoss) [reply packages/packages/documents/src/init.ts:26]: Done

File: packages/packages/site-settings/src/components/__tests__/ported-primary-action.test.tsx (hunk 3 above)
Comment (Nevoss) [reply packages/packages/site-settings/src/components/__tests__/ported-primary-action.test.tsx:93]: For another task

File: packages/packages/site-settings/src/components/__tests__/ported-primary-action.test.tsx (hunk 4 above)
Comment (Nevoss) [reply packages/packages/site-settings/src/components/__tests__/ported-primary-action.test.tsx:30]: Dibe

File: packages/packages/site-settings/src/components/__tests__/ported-primary-action.test.tsx (hunk 5 above)
Comment (Nevoss) [reply packages/packages/site-settings/src/components/__tests__/ported-primary-action.test.tsx:45]: No

File: packages/packages/site-settings/src/components/__tests__/ported-primary-action.test.tsx (hunk 6 above)
Comment (Nevoss) [reply packages/packages/site-settings/src/components/__tests__/ported-primary-action.test.tsx:48]: Done

File: packages/packages/site-settings/src/components/__tests__/ported-primary-action.test.tsx (hunk 7 above)
Comment (Nevoss) [reply packages/packages/site-settings/src/components/__tests__/ported-primary-action.test.tsx:60]: No

File: packages/packages/site-settings/src/components/__tests__/ported-primary-action.test.tsx (hunk 8 above)
Comment (Nevoss) [reply packages/packages/site-settings/src/components/__tests__/ported-primary-action.test.tsx:68]: Done

File: packages/packages/site-settings/src/components/__tests__/ported-primary-action.test.tsx (hunk 9 above)
Comment (Nevoss) [reply packages/packages/site-settings/src/components/__tests__/ported-primary-action.test.tsx:19]: Done

File: packages/packages/site-settings/src/components/__tests__/primary-action.test.tsx (hunk 10 above)
Comment (Nevoss) [reply packages/packages/site-settings/src/components/__tests__/primary-action.test.tsx:11]: Done

File: packages/packages/site-settings/src/components/__tests__/primary-action.test.tsx (hunk 11 above)
Comment (Nevoss) [reply packages/packages/site-settings/src/components/__tests__/primary-action.test.tsx:54]: Done

File: packages/packages/site-settings/src/components/__tests__/primary-action.test.tsx (hunk 13 above)
Comment (Nevoss) [reply packages/packages/site-settings/src/components/__tests__/primary-action.test.tsx:76]: Done

File: packages/packages/site-settings/src/components/__tests__/primary-action.test.tsx (hunk 14 above)
Comment (Nevoss) [reply packages/packages/site-settings/src/components/__tests__/primary-action.test.tsx:69]: Done

File: packages/packages/site-settings/src/components/__tests__/primary-action.test.tsx (hunk 15 above)
Comment (Nevoss) [reply packages/packages/site-settings/src/components/__tests__/primary-action.test.tsx:73]: Done

File: packages/packages/site-settings/src/components/__tests__/primary-action.test.tsx (hunk 16 above)
Comment (Nevoss) [reply packages/packages/site-settings/src/components/__tests__/primary-action.test.tsx:86]: Done

File: packages/packages/site-settings/src/components/portal.tsx (hunk 17 above)
Comment (Nevoss) [reply packages/packages/site-settings/src/components/portal.tsx:13]: Done

File: packages/packages/site-settings/src/components/ported-primary-action.tsx (hunk 18 above)
Comment (Nevoss) [reply packages/packages/site-settings/src/components/ported-primary-action.tsx:4]: Done

File: packages/packages/site-settings/src/components/primary-action.tsx (hunk 19 above)
Comment (Nevoss) [reply packages/packages/site-settings/src/components/primary-action.tsx:23]: No

File: packages/packages/site-settings/src/components/portal.tsx (hunk 20 above)
Comment (Nevoss) [reply packages/packages/site-settings/src/components/portal.tsx:5]: Done

File: packages/packages/site-settings/src/hooks/__tests__/use-action-props.test.tsx (hunk 21 above)
Comment (Nevoss) [reply packages/packages/site-settings/src/hooks/__tests__/use-action-props.test.tsx:17]: Done

File: packages/packages/site-settings/src/hooks/__tests__/use-action-props.test.tsx (hunk 22 above)
Comment (Nevoss) [reply packages/packages/site-settings/src/hooks/__tests__/use-action-props.test.tsx:18]: Done

File: packages/packages/site-settings/src/hooks/use-action-props.tsx (hunk 24 above)
Comment (Nevoss) [reply packages/packages/site-settings/src/hooks/use-action-props.tsx:1]: Done

File: packages/packages/site-settings/src/hooks/use-action-props.tsx (hunk 25 above)
Comment (Nevoss) [reply packages/packages/site-settings/src/hooks/use-action-props.tsx:21]: Done

**Review by StyleShit — DISMISSED** [review]

File: packages/packages/site-settings/src/components/primary-action.tsx [hunk 27]
```diff
@@ -0,0 +1,36 @@
...
+			position: 'absolute',
+			bottom: 0,
+			right: 0,
+			left: 0,
+			px: 5,
+			py: 4,
+			borderTop: 1,
+			borderColor: 'grey.200',
+			background: 'white',
+		} }>
```
Comment (StyleShit) [inline packages/packages/site-settings/src/components/primary-action.tsx:20]: Paper,
remove background,
broderColor: 'divider'

**Review by rotemee — CHANGES_REQUESTED** [review]

File: packages/packages/site-settings/src/components/portal.tsx [hunk 28]
```diff
@@ -0,0 +1,29 @@
+import { ReactNode } from 'react';
+import { Portal as BasePortal } from '@elementor/ui';
+import { isRouteActive, routeCloseEvent, routeOpenEvent, useListenTo } from '@elementor/v1-adapters';
+
+export default function Portal( { children }: { children: ReactNode } ) {
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply packages/packages/site-settings/src/components/portal.tsx:5]: Done

File: packages/packages/site-settings/src/icons/settings-icon.tsx [hunk 29]
```diff
@@ -0,0 +1,9 @@
+import { SvgIcon } from '@elementor/ui';
+
+export default function SettingsIcon() {
+	return (
+		<SvgIcon viewBox="0 0 16 16">
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply packages/packages/site-settings/src/icons/settings-icon.tsx]: Done

**Review by rotemee — APPROVED** [review]
